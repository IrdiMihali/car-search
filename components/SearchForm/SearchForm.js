import { useEffect, useState, useRef } from 'react';
import PropTypes from 'prop-types';

import { useDebounce } from '../../hooks/useDebounce';
import { usePopup } from '../../hooks/usePopup';
import {
	StyledForm,
	Type,
	SearchButton,
	List,
	Spinner,
	StyledDropdown,
	StyledListItem,
	StyledInformation,
	StyledDescription,
	StyledInput,
} from './SearchForm.styled';

const MAX_RESULTS_INDEX = 5;

const SearchForm = ({ resultList, onSearch, loading }) => {
	const [inputValue, setInputValue] = useState('');
	const [searchQuery, setSearchQuery] = useState('');
	const [selectedItemIdx, setSelectedItemIdx] = useState();
	const debouncedSearchQuery = useDebounce(searchQuery, 100);
	const ref = useRef();
	const [isListOpen, setIsListOpen] = usePopup(!!resultList.length, ref);

	useEffect(() => {
		onSearch(debouncedSearchQuery);
	}, [debouncedSearchQuery]);

	const onInputChange = (inputValue) => {
		setSearchQuery(inputValue);
		setInputValue(inputValue);
		setIsListOpen(true);
	};
	const onListItemSelect = (value, index) => {
		setSearchQuery('');
		setInputValue(value);
		setSelectedItemIdx(index);
		setIsListOpen(false);
	};
	const onSubmit = (ev) => ev.preventDefault();
	const onKeyDown = (evt) => {
		if (!resultList || !isListOpen) return;

		switch (evt.key) {
			case 'ArrowDown':
				evt.preventDefault();

				if (selectedItemIdx === undefined) {
					setSelectedItemIdx(0);
				} else {
					selectedItemIdx < MAX_RESULTS_INDEX
						? setSelectedItemIdx(selectedItemIdx + 1)
						: setSelectedItemIdx(MAX_RESULTS_INDEX);
				}
				break;

			case 'ArrowUp':
				evt.preventDefault();

				if (selectedItemIdx === undefined) return;
				else {
					selectedItemIdx > 0
						? setSelectedItemIdx(selectedItemIdx - 1)
						: setSelectedItemIdx(0);
				}
				break;

			case 'Enter':
				evt.preventDefault();

				if (
					selectedItemIdx === undefined ||
					resultList[selectedItemIdx] === undefined
				)
					return;

				onListItemSelect(resultList[selectedItemIdx].value, selectedItemIdx);

				break;

			case 'Escape':
				evt.preventDefault();

				setIsListOpen(false);

				break;

			default:
				break;
		}
	};

	return (
		<form onSubmit={onSubmit}>
			<StyledForm>
				<div
					data-testid="results-combobox"
					role="combobox"
					aria-haspopup={resultList && `listbox`}
					aria-expanded={resultList && isListOpen}
					aria-activedescendant={resultList && isListOpen && selectedItemIdx}
				>
					<svg viewBox="0 0 24 24" width="1em" height="1em">
						<path d="M17.464 6.56a8.313 8.313 0 1 1-15.302 6.504A8.313 8.313 0 0 1 17.464 6.56zm1.38-.586C16.724.986 10.963-1.339 5.974.781.988 2.9-1.337 8.662.783 13.65c2.12 4.987 7.881 7.312 12.87 5.192 4.987-2.12 7.312-7.881 5.192-12.87zM15.691 16.75l7.029 7.03a.75.75 0 0 0 1.06-1.06l-7.029-7.03a.75.75 0 0 0-1.06 1.06z"></path>
					</svg>
					<StyledInput
						aria-controls={resultList && 'location-search-listbox'}
						aria-autocomplete={resultList}
						onKeyDown={onKeyDown}
						aria-label="Pick-up location"
						data-testid="search-input"
						id="location-search"
						name="q"
						onChange={({ target }) => onInputChange(target.value)}
						placeholder="Pick-up location"
						type="search"
						value={inputValue}
					/>
					{loading && (
						<Spinner data-testid="loading-spinner">
							<i />
						</Spinner>
					)}
				</div>
				<SearchButton type="submit" shouldShow={isListOpen}>
					Search
				</SearchButton>
			</StyledForm>
			<StyledDropdown>
				{isListOpen && (
					<List
						data-testid="results-list"
						id="location-search-listbox"
						ref={ref}
						role="listbox"
					>
						{resultList.length > 0 &&
							resultList.map(
								({ type, title, description, id, value }, index) => {
									const isFocused = index === selectedItemIdx;
									return (
										<StyledListItem
											aria-hidden={!isListOpen}
											aria-selected={isFocused}
											data-testid={id}
											id={id}
											isFocused={isFocused}
											key={id}
											onClick={() => onListItemSelect(value, index)}
											role="option"
										>
											<Type type={type}>{type}</Type>
											<StyledInformation>
												<div>{title}</div>
												<StyledDescription>{description}</StyledDescription>
											</StyledInformation>
										</StyledListItem>
									);
								}
							)}
					</List>
				)}
			</StyledDropdown>
		</form>
	);
};

SearchForm.propTypes = {
	resultList: PropTypes.arrayOf(
		PropTypes.shape({
			id: PropTypes.string.isRequired,
			type: PropTypes.string,
			title: PropTypes.string.isRequired,
			description: PropTypes.string,
			value: PropTypes.string.isRequired,
		})
	),
	onSearch: PropTypes.func.isRequired,
	loading: PropTypes.bool.isRequired,
};

export default SearchForm;
