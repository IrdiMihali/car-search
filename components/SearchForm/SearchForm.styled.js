import styled, { css } from 'styled-components';

export const StyledForm = styled.div`
	display: flex;
	flex-direction: column;
	justify-content: space-between;
	max-width: 1142px;
	padding: 0 16px;
	position: relative;

	@media (min-width: 1024px) {
		padding-top: 16px;
		flex-direction: row;
	}

	> div:first-of-type {
		align-items: center;
		display: flex;
		display: inline-flex;
		height: var(--input-height);
		position: relative;
		width: 100%;

		svg {
			fill: #000;
			height: 16px;
			left: 16px;
			pointer-events: none;
			position: absolute;
			right: auto;
			width: 16px;
			z-index: 1;
		}
	}
`;

export const StyledInput = styled.input`
	border-radius: 10px 10px 0px 0px;
	border: 4px solid var(--main-border-color);
	box-sizing: border-box;
	font-size: 1em;
	font-weight: 400;
	height: 100%;
	left: 0;
	padding: 25px 40px 24px 40px;
	top: 0;
	width: 100%;

	:focus {
		outline-width: 0;
		::placeholder {
			transform: translateY(-1.5em);
			font-size: 0.7em;
		}
	}

	:hover {
		outline-width: 0;
	}

	::-webkit-search-cancel-button,
	::-webkit-search-results-decoration {
		display: none;
	}

	@media (min-width: 1024px) {
		border-radius: 10px 0px 0px 10px;
	}
`;

export const Spinner = styled.div`
	position: absolute;
	right: 10px;
	top: calc(50% - 10px);

	i {
		animation: around 5.4s infinite;
		display: inline-block;
		height: 20px;
		position: relative;
		width: 20px;
	}

	@keyframes around {
		0% {
			transform: rotate(0deg);
		}
		100% {
			transform: rotate(360deg);
		}
	}

	i::after,
	i::before {
		content: '';
		animation: around 0.7s ease-in-out infinite;
		border-color: #1879ca #1879ca transparent transparent;
		border-radius: 20px;
		border-style: solid;
		border-width: 2px;
		box-sizing: border-box;
		display: inline-block;
		height: 100%;
		position: absolute;
		width: 100%;
	}

	i::after {
		animation: around 0.7s ease-in-out 0.1s infinite;
		background: transparent;
	}
`;

export const StyledDropdown = styled.div`
	display: flex;
	flex-direction: column;
	justify-content: space-between;
	max-width: 1142px;
	padding: 0 16px;
	position: relative;

	@media (min-width: 1024px) {
		flex-direction: row;
	}
`;

export const List = styled.ul`
	animation: fadeIn 0.5s;
	background-color: #fff;
	border-radius: 4px;
	border: none;
	box-shadow: 0 2px 8px rgba(0, 0, 0, 0.16);
	margin-block-end: 0;
	margin-block-start: 0;
	max-height: 100vh;
	overflow-y: scroll;
	padding-inline-start: 0;
	padding: 0px;
	position: relative;
	transition: max-height 0.5s ease-in-out, opacity 0.1s ease;
	width: 100%;
	will-change: opacity;
`;

export const StyledListItem = styled.li`
	align-items: center;
	border-radius: 4px;
	cursor: pointer;
	display: flex;
	flex-wrap: nowrap;
	gap: 10px;
	padding: calc(4px * 3);
	text-align: match-parent;

	${({ isFocused }) =>
		isFocused &&
		css`
			background-color: #d1e8fa;
		`}
`;

export const Type = styled.div`
	align-items: center;
	border-radius: 4px;
	border: none;
	color: #fff;
	display: inline-block;
	display: inline-flex;
	flex-shrink: 0;
	font-size: 0.8em;
	font-weight: 500;
	height: 0.8rem;
	justify-content: center;
	line-height: 0.8rem;
	min-width: 5.1rem;
	padding: 0.3rem 0;
	padding: 8px;
	text-align: center;
	width: 70px;

	${({ type }) =>
		type === 'Airport'
			? css`
					color: #262626;
					background: #ff8000;
			  `
			: type === 'City' || type === 'Station'
			? css`
					background: #1879ca;
			  `
			: type === 'Region'
			? css`
					color: #262626;
					background: #f1c74c;
			  `
			: type === 'Region'
			? css`
					color: #262626;
					background: #068323;
			  `
			: css`
					display: none;
			  `}
`;

export const StyledInformation = styled.div`
	align-items: flex-start;
	display: flex;
	flex-direction: column;
	font-weight: 400;
	line-height: 20px;
	text-align: start;
`;
export const StyledDescription = styled.div`
	font-size: 0.8em;
	color: #474747;
	font-weight: 500;
`;

export const SearchButton = styled.button`
	align-items: center;
	background-color: #068323 !important;
	border-bottom: 4px solid var(--main-border-color);
	border-left: 4px solid var(--main-border-color);
	border-radius: 0px 0px 10px 10px;
	border-right: 4px solid var(--main-border-color);
	border-top: none;
	box-sizing: border-box;
	color: #fff;
	cursor: pointer;
	font-size: 1.4em;
	font-weight: 700;
	height: var(--input-height);
	justify-content: center;
	line-height: normal;
	margin: 0;
	min-height: 44px;
	overflow: visible;
	padding: calc(4px * 2) 16px;
	text-align: center;
	text-decoration: none;
	transition: background-color ease-in-out 0.1s, color ease-in-out 0.1s;
	vertical-align: middle;
	display: ${({ shouldShow }) => shouldShow && 'none'};

	:active {
		background-color: #056b1d;
	}

	@media (min-width: 1024px) {
		border-bottom: 4px solid var(--main-border-color);
		border-left: none;
		border-radius: 0px 10px 10px 0px;
		border-right: 4px solid var(--main-border-color);
		border-top: 4px solid var(--main-border-color);
		display: block;
	}
`;
