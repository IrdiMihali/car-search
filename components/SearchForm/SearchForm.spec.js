import { render, fireEvent, screen } from '@testing-library/react';
import { act } from 'react-dom/test-utils';
import 'jest-styled-components';

import SearchForm from './SearchForm';

const resultListMock = [
	{
		description: 'Bergamo, Lombardy, Italy',
		id: 'airport-19926',
		title: 'Bergamo Airport (BGY) ',
		type: 'Airport',
		value: 'Bergamo Airport (BGY) , Bergamo, Lombardy, Italy',
	},
	{
		description: 'Lombardia, Italy',
		id: 'city-1295131',
		title: 'Milan',
		type: 'City',
		value: 'Milan, Lombardia, Italy',
	},
	{
		description: 'Lombardia, Italy',
		id: 'district-1295131',
		title: 'Milan',
		type: 'District',
		value: 'Milan, Lombardia, Italy',
	},
	{
		description: 'Lombardia, Italy',
		id: 'station-1295131',
		title: 'Milan',
		type: 'Station',
		value: 'Milan, Lombardia, Italy',
	},
];

describe('Given `SearchForm` component', () => {
	const initialProps = {
		resultList: [],
		onSearch: jest.fn(),
		loading: false,
	};

	jest.useFakeTimers();

	it('should render successfully', () => {
		const { container } = render(<SearchForm {...initialProps} />);

		expect(container).toMatchSnapshot();
	});

	describe('When the user types in the input', () => {
		it('should show a spinner while it is loading', () => {
			render(<SearchForm {...initialProps} loading />);

			fireEvent.change(screen.getByTestId('search-input'), {
				target: { value: 'Italy' },
			});

			expect(screen.queryByTestId('loading-spinner')).toBeInTheDocument();
		});

		it('should trigger the `onSearch` function', () => {
			const onSearch = jest.fn();
			render(<SearchForm {...initialProps} onSearch={onSearch} />);

			fireEvent.change(screen.getByTestId('search-input'), {
				target: { value: 'Italy' },
			});

			act(() => jest.runAllTimers());

			expect(onSearch).toHaveBeenCalledWith('Italy');
		});

		describe('and the user enters 1 character', () => {
			it('should NOT display the listbox', () => {
				render(<SearchForm {...initialProps} />);
				const listbox = screen.queryByTestId('results-list');

				fireEvent.change(screen.getByTestId('search-input'), {
					target: { value: 'I' },
				});

				expect(listbox).not.toBeInTheDocument();
			});
		});

		describe('and the user enters more than 1 character', () => {
			it('should display the list of results', () => {
				render(<SearchForm {...initialProps} resultList={resultListMock} />);
				const input = screen.getByTestId('search-input');
				const listbox = screen.queryByTestId('results-list');

				fireEvent.change(input, { target: { value: 'Ital' } });

				expect(listbox).toBeInTheDocument();
			});

			describe('and the user presses the Arrow Down Key', () => {
				beforeEach(() => {
					render(<SearchForm {...initialProps} resultList={resultListMock} />);
				});

				it('should focus the first option in the list', () => {
					fireEvent.keyDown(screen.getByTestId('search-input'), {
						key: 'ArrowDown',
					});

					expect(screen.queryByTestId(resultListMock[0].id)).toHaveAttribute(
						'aria-selected',
						'true'
					);
				});
			});

			describe('and the user presses the Arrow Up Key from the forth option', () => {
				beforeEach(() => {
					render(<SearchForm {...initialProps} resultList={resultListMock} />);
					resultListMock.forEach(() =>
						fireEvent.keyDown(screen.getByTestId('search-input'), {
							key: 'ArrowDown',
						})
					);
				});

				it('should focus the third option in the list', () => {
					fireEvent.keyDown(screen.getByTestId('search-input'), {
						key: 'ArrowUp',
					});

					expect(screen.queryByTestId(resultListMock[2].id)).toHaveAttribute(
						'aria-selected',
						'true'
					);
				});
			});

			describe('and the user presses the Enter key', () => {
				beforeEach(() => {
					render(<SearchForm {...initialProps} resultList={resultListMock} />);
					fireEvent.keyDown(screen.getByTestId('search-input'), {
						key: 'ArrowDown',
					});
				});

				it('should close the dropdown list and select the item', () => {
					fireEvent.keyDown(screen.getByTestId('search-input'), {
						key: 'Enter',
					});

					expect(screen.queryByTestId('results-list')).not.toBeInTheDocument();

					expect(screen.getByTestId('search-input').value).toBe(
						resultListMock[0].value
					);
				});
			});

			describe('and the user clicks the listbox option', () => {
				it('should select the listbox option', () => {
					render(<SearchForm {...initialProps} resultList={resultListMock} />);
					const listItem = screen.queryByTestId(resultListMock[0].id);

					fireEvent.click(listItem);

					expect(screen.getByTestId('search-input').value).toBe(
						resultListMock[0].value
					);
				});
			});
		});
	});
});
