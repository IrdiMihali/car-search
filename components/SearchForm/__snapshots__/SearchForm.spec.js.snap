// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given \`SearchForm\` component should render successfully 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  max-width: 1142px;
  padding: 0 16px;
  position: relative;
}

.c0 > div:first-of-type {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  height: var(--input-height);
  position: relative;
  width: 100%;
}

.c0 > div:first-of-type svg {
  fill: #000;
  height: 16px;
  left: 16px;
  pointer-events: none;
  position: absolute;
  right: auto;
  width: 16px;
  z-index: 1;
}

.c1 {
  border-radius: 10px 10px 0px 0px;
  border: 4px solid var(--main-border-color);
  box-sizing: border-box;
  font-size: 1em;
  font-weight: 400;
  height: 100%;
  left: 0;
  padding: 25px 40px 24px 40px;
  top: 0;
  width: 100%;
}

.c1:focus {
  outline-width: 0;
}

.c1:focus::-webkit-input-placeholder {
  -webkit-transform: translateY(-1.5em);
  -ms-transform: translateY(-1.5em);
  transform: translateY(-1.5em);
  font-size: 0.7em;
}

.c1:focus::-moz-placeholder {
  -webkit-transform: translateY(-1.5em);
  -ms-transform: translateY(-1.5em);
  transform: translateY(-1.5em);
  font-size: 0.7em;
}

.c1:focus:-ms-input-placeholder {
  -webkit-transform: translateY(-1.5em);
  -ms-transform: translateY(-1.5em);
  transform: translateY(-1.5em);
  font-size: 0.7em;
}

.c1:focus::placeholder {
  -webkit-transform: translateY(-1.5em);
  -ms-transform: translateY(-1.5em);
  transform: translateY(-1.5em);
  font-size: 0.7em;
}

.c1:hover {
  outline-width: 0;
}

.c1::-webkit-search-cancel-button,
.c1::-webkit-search-results-decoration {
  display: none;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  max-width: 1142px;
  padding: 0 16px;
  position: relative;
}

.c2 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #068323 !important;
  border-bottom: 4px solid var(--main-border-color);
  border-left: 4px solid var(--main-border-color);
  border-radius: 0px 0px 10px 10px;
  border-right: 4px solid var(--main-border-color);
  border-top: none;
  box-sizing: border-box;
  color: #fff;
  cursor: pointer;
  font-size: 1.4em;
  font-weight: 700;
  height: var(--input-height);
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  line-height: normal;
  margin: 0;
  min-height: 44px;
  overflow: visible;
  padding: calc(4px * 2) 16px;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-transition: background-color ease-in-out 0.1s,color ease-in-out 0.1s;
  transition: background-color ease-in-out 0.1s,color ease-in-out 0.1s;
  vertical-align: middle;
}

.c2:active {
  background-color: #056b1d;
}

@media (min-width:1024px) {
  .c0 {
    padding-top: 16px;
    -webkit-flex-direction: row;
    -ms-flex-direction: row;
    flex-direction: row;
  }
}

@media (min-width:1024px) {
  .c1 {
    border-radius: 10px 0px 0px 10px;
  }
}

@media (min-width:1024px) {
  .c3 {
    -webkit-flex-direction: row;
    -ms-flex-direction: row;
    flex-direction: row;
  }
}

@media (min-width:1024px) {
  .c2 {
    border-bottom: 4px solid var(--main-border-color);
    border-left: none;
    border-radius: 0px 10px 10px 0px;
    border-right: 4px solid var(--main-border-color);
    border-top: 4px solid var(--main-border-color);
    display: block;
  }
}

<div>
  <form>
    <div
      class="c0"
    >
      <div
        aria-activedescendant="false"
        aria-expanded="false"
        aria-haspopup="listbox"
        data-testid="results-combobox"
        role="combobox"
      >
        <svg
          height="1em"
          viewBox="0 0 24 24"
          width="1em"
        >
          <path
            d="M17.464 6.56a8.313 8.313 0 1 1-15.302 6.504A8.313 8.313 0 0 1 17.464 6.56zm1.38-.586C16.724.986 10.963-1.339 5.974.781.988 2.9-1.337 8.662.783 13.65c2.12 4.987 7.881 7.312 12.87 5.192 4.987-2.12 7.312-7.881 5.192-12.87zM15.691 16.75l7.029 7.03a.75.75 0 0 0 1.06-1.06l-7.029-7.03a.75.75 0 0 0-1.06 1.06z"
          />
        </svg>
        <input
          aria-autocomplete=""
          aria-controls="location-search-listbox"
          aria-label="Pick-up location"
          class="c1"
          data-testid="search-input"
          id="location-search"
          name="q"
          placeholder="Pick-up location"
          type="search"
          value=""
        />
      </div>
      <button
        class="c2"
        type="submit"
      >
        Search
      </button>
    </div>
    <div
      class="c3"
    />
  </form>
</div>
`;
